[dynamo] trace local_map with local shapes for AP

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164437
* #164433
* #164431
* __->__ #163602
* #164340
* #164420
* #164419
* #164321
* #164296

Context is in https://www.internalfb.com/excalidraw/EX519691 and https://docs.google.com/document/d/1qnuXLZk_GYt_PksHTwkn7L2ELRDnYlIRPkHAlXTyuhw/edit?tab=t.0. And the description of the previous PR: https://github.com/pytorch/pytorch/pull/164340.

The previous PR adds the support on the HOP side for eager execution and AOTAutograd. Dynamo is still passing the HOP a subgraph with wrong shapes. This PR fixes that. This is similar to the HOP implementation, however we additionally need to manually keep the TensorVariable metadata in sync.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

xmfan committed:
[hop] trace local_map with local shapes for AP

[ghstack-poisoned]

xmfan committed:
Update on "[hop] trace local_map with local shapes for AP"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xmfan committed:
Update on "[hop] trace local_map with local shapes for AP"


So the eager semantics of local_map are to no-op plain tensor inputs, and to redistribute DTensor inputs.

In AP, we will do the initial trace with global shapes and shard them later. This means that prior to this PR, we are incorrectly tracing the local_map body with global shapes. This can cause problems if we specialize on shapes e.g. view ops, because we would then error later when we trace again with local shapes post sharding.

The outcome of the discussion in https://docs.google.com/document/d/1qnuXLZk_GYt_PksHTwkn7L2ELRDnYlIRPkHAlXTyuhw/edit?tab=t.0 was that we would trace the local_map body with local shapes (redistribute them as if they were DTensors).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xmfan committed:
Update on "[hop] trace local_map with local shapes for AP"


So the eager semantics of local_map are to no-op plain tensor inputs, and to redistribute DTensor inputs.

In AP, we will do the initial trace with global shapes and shard them later. This means that prior to this PR, we are incorrectly tracing the local_map body with global shapes. This can cause problems if we specialize on shapes e.g. view ops, because we would then error later when we trace again with local shapes post sharding.

The outcome of the discussion in https://docs.google.com/document/d/1qnuXLZk_GYt_PksHTwkn7L2ELRDnYlIRPkHAlXTyuhw/edit?tab=t.0 was that we would trace the local_map body with local shapes (redistribute them as if they were DTensors).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xmfan committed:
Update on "[hop] trace local_map with local shapes for AP"


So the eager semantics of local_map are to no-op plain tensor inputs, and to redistribute DTensor inputs.

In AP, we will do the initial trace with global shapes and shard them later. This means that prior to this PR, we are incorrectly tracing the local_map body with global shapes. This can cause problems if we specialize on shapes e.g. view ops, because we would then error later when we trace again with local shapes post sharding.

The outcome of the discussion in https://docs.google.com/document/d/1qnuXLZk_GYt_PksHTwkn7L2ELRDnYlIRPkHAlXTyuhw/edit?tab=t.0 was that we would trace the local_map body with local shapes (redistribute them as if they were DTensors).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xmfan committed:
Update on "[dynamo] trace local_map with local shapes for AP"


So the eager semantics of local_map are to no-op plain tensor inputs, and to redistribute DTensor inputs.

In AP, we will do the initial trace with global shapes and shard them later. This means that prior to this PR, we are incorrectly tracing the local_map body with global shapes. This can cause problems if we specialize on shapes e.g. view ops, because we would then error later when we trace again with local shapes post sharding.

The outcome of the discussion in https://docs.google.com/document/d/1qnuXLZk_GYt_PksHTwkn7L2ELRDnYlIRPkHAlXTyuhw/edit?tab=t.0 was that we would trace the local_map body with local shapes (redistribute them as if they were DTensors).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xmfan committed:
Update on "[dynamo] trace local_map with local shapes for AP"


So the eager semantics of local_map are to no-op plain tensor inputs, and to redistribute DTensor inputs.

In AP, we will do the initial trace with global shapes and shard them later. This means that prior to this PR, we are incorrectly tracing the local_map body with global shapes. This can cause problems if we specialize on shapes e.g. view ops, because we would then error later when we trace again with local shapes post sharding.

The outcome of the discussion in https://docs.google.com/document/d/1qnuXLZk_GYt_PksHTwkn7L2ELRDnYlIRPkHAlXTyuhw/edit?tab=t.0 was that we would trace the local_map body with local shapes (redistribute them as if they were DTensors).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

ezyang commented:
still draft?

xmfan committed:
Update on "[dynamo] trace local_map with local shapes for AP"


So the eager semantics of local_map are to no-op plain tensor inputs, and to redistribute DTensor inputs.

In AP, we will do the initial trace with global shapes and shard them later. This means that prior to this PR, we are incorrectly tracing the local_map body with global shapes. This can cause problems if we specialize on shapes e.g. view ops, because we would then error later when we trace again with local shapes post sharding.

The outcome of the discussion in https://docs.google.com/document/d/1qnuXLZk_GYt_PksHTwkn7L2ELRDnYlIRPkHAlXTyuhw/edit?tab=t.0 was that we would trace the local_map body with local shapes (redistribute them as if they were DTensors).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

xmfan committed:
Update on "[dynamo] trace local_map with local shapes for AP"


So the eager semantics of local_map are to no-op plain tensor inputs, and to redistribute DTensor inputs.

In AP, we will do the initial trace with global shapes and shard them later. This means that prior to this PR, we are incorrectly tracing the local_map body with global shapes. This can cause problems if we specialize on shapes e.g. view ops, because we would then error later when we trace again with local shapes post sharding.

The outcome of the discussion in https://docs.google.com/document/d/1qnuXLZk_GYt_PksHTwkn7L2ELRDnYlIRPkHAlXTyuhw/edit?tab=t.0 was that we would trace the local_map body with local shapes (redistribute them as if they were DTensors).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -3576,6 +3576,63 @@ def _call_function(
+            local_tensor = local_tensor[0]
+
+            priors[vt] = global_tensor
+            vt.as_proxy().node.meta["example_value"] = local_tensor
```
Mutating an existing variable tracker looks suspicious as some other subsystems might assume the mutation always come from interpreting user code that actually does the mutation. There could be some future risk if we don't differentiate these two cases e.g. these two syncs might get recorded as two attribute mutations.

The alternative is to create new tensor variables with desirable shapes e.g.like in [flex_attention](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/higher_order_ops.py#L2844-L2850), we create new_empty tensors and enclosing them with discard_graph_changes().

xmfan committed:
Update on "[dynamo] trace local_map with local shapes for AP"


Context is in https://www.internalfb.com/excalidraw/EX519691 and https://docs.google.com/document/d/1qnuXLZk_GYt_PksHTwkn7L2ELRDnYlIRPkHAlXTyuhw/edit?tab=t.0. And the description of the previous PR: https://github.com/pytorch/pytorch/pull/164340.

The previous PR adds the support on the HOP side for eager execution and AOTAutograd. Dynamo is still passing the HOP a subgraph with wrong shapes. This PR fixes that. This is similar to the HOP implementation, however we additionally need to manually keep the TensorVariable metadata in sync.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]
